### PR TITLE
Block Bindings: Update label for bound paragraphs and headings

### LINKED ISSUE #62933: Block Bindings: Improve accessibility of bound paragraphs when using a screenreader in Navigation Mode
### Description

Actual: Currently, when tabbing to a bound paragraph in Navigation Mode, the NVDA screenreader announces the bound paragraph's original content, not its override content, which may be confusing and giver users incorrect information of what is displayed on the page.

Expected: The screenreader announces the override content, not the original content.

### Step-by-step reproduction instructions

<details>
<summary>1. Register post meta by adding this snippet to your theme's functions.php</summary>

```php
add_action( 'init', 'test_block_bindings' );

function test_block_bindings() {
	register_meta(
		'post',
		'text_field',
		array(
			'show_in_rest'      => true,
			'single'            => true,
			'type'              => 'string',
			'default'           => 'default text value',
		)
	);
}
```

</details>

<details>

<summary>2. Add a paragraph block bound to the custom field using the Code Editor</summary>

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_field"}}}}} -->
<p>Bound paragraph</p>
<!-- /wp:paragraph -->
```

</details>
 
3. Override the paragraph content by typing into the bound paragraph.

4. Using the NVDA screen reader, navigate to the bound paragraph in Navigation Mode, and see that the paragraph's original content, rather than its bound content, is announced.

### Screenshots, screen recording, code snippet

(Pardon the background noise — I was unable to use my isolated microphone to capture the sound of the screenreader)

https://github.com/WordPress/gutenberg/assets/5360536/f8428579-3b69-4579-be3c-e5707363eb1c

### Environment info

- Testing using NVDA screen reader on Windows Firefox
- WP 6.5, Gutenberg 18.7.0-rc.1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by artemiomorales**
## What?

This PR fixes the label for bound paragraphs and headings, using their bindings value rather than fallback content.

## Why?

Addresses https://github.com/WordPress/gutenberg/issues/62933
The label is currently read out loud by screen readers. Unless we update the label for bound blocks, screen readers will simply announce the value in the markup, which is misleading, as the actual content will be rendered by the block bindings API.

## How?

It replicates some logic from `use-bindings-attributes.js` inside of `block-selection-button.js` in order to retrieve the bindings value, and also extracts key logic into a new `getBindingsValues` function so that it can be reused.

## Testing Instructions

<details>
<summary>1. Register post meta by adding this snippet to your theme's functions.php</summary>

```php
add_action( 'init', 'test_block_bindings' );

function test_block_bindings() {
	register_meta(
		'post',
		'text_field',
		array(
			'show_in_rest'      => true,
			'single'            => true,
			'type'              => 'string',
			'default'           => 'default text value',
		)
	);
}
```

</details>

<details>

<summary>2. Add heading and paragraph blocks bound to a custom field using the Code Editor</summary>

```
<!-- wp:heading {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_field"}}}}} -->
<h2 class="wp-block-heading">Bound heading</h2>
<!-- /wp:heading -->
```

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_field"}}}}} -->
<p>Bound paragraph</p>
<!-- /wp:paragraph -->
```
</details>
 
3. Using the NVDA screen reader, navigate to the bound blocks in Navigation Mode, and see that the their modified content is announced.

### CODE REVIEW

**Comment by SantosGuillamot** [comment]
I assume this is going to be the case for other blocks and attributes. For example, does this happen with the heading content or the button text?

If that's the case, I believe we probably need to think about it more broadly and solve it for any block attribute, potentially using the bound value instead of a default message.

**Comment by gziolo** [comment]
If accepted, it should be replicated with other blocks that can be connected. The challenge is that the attribute connected isn'tand always the most important one,  connections for attributes editable in the sidebar shouldn't become too prominent.

**Comment by artemiomorales** [comment]
To start, I just hardcoded a new label for bound paragraphs to get the conversation started.

Ideally, we would retrieve the actual binding value, but I'm not sure what would be the best way to go about that. The logic for reading the bindings value is currently in [use-binding-attributes.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/use-bindings-attributes.js), and I'm not sure it's necessary to get so involved.

In addition to some text indicating that the content is dynamic, perhaps even just using the binding source and key could suffice? That information is already easily available in the block attributes.

**Comment by SantosGuillamot** [comment]
Personally, I don't feel comfortable hardcoding something that just works for the paragraph. If it is the `__experimentalLabel` setting what we need to change, maybe we can filter that with the existing hook or a new one. I haven't looked into that.

**Comment by artemiomorales** [comment]
> I assume this is going to be the case for other blocks and attributes. For example, does this happen with the heading content or the button text?

@SantosGuillamot This indeed happens with the heading block.
We don't currently announce the text content of button blocks in navigation mode, though.

> If that's the case, I believe we probably need to think about it more broadly and solve it for any block attribute, potentially using the bound value instead of a default message.

If it's just the paragraph and heading, do you think that warrants a broader solution?

Potentially we could intervene at an earlier level. For example, this is where the `__experimentalLabel` gets read. Potentially we could check for bindings at this stage and modify the label accordingly:

https://github.com/WordPress/gutenberg/blob/1a8aa20d057e6bd623d3ff8178b2d67de2845dcc/packages/blocks/src/api/utils.js#L148-L173

**Comment by SantosGuillamot** [comment]
> We don't currently announce the text content of button blocks in navigation mode, though.
> If it's just the paragraph and heading, do you think that warrants a broader solution?

Is it expected or it is just because we aren't passing `__experimentalLabel` to the button block?

I'm not familiar with this part of the codebase, but I guess we will encounter this problem for any block using `__experimentalLabel`, so I'd try to fix the root of the problem, not adding a workaround for the use cases we have identified.

For example, it seems it is also used in the image block: [link](https://github.com/WordPress/gutenberg/blob/322766e04ab3351df9cbf5ee22c220ff03a72d38/packages/block-library/src/image/index.js#L31). I haven't tested it, but maybe there is the same problem there.

There are already a bunch of blocks relying on that, and it could even create more issues when we add support for more blocks.

Looking at the `getBlockLabel` and `getBlockAccessibilityLabel` seems like a good starting point.

**Comment by artemiomorales** [comment]
> Is it expected or it is just because we aren't passing __experimentalLabel to the button block?

I believe it's because __experimentalLabel hasn't been defined for the button block.

> I'm not familiar with this part of the codebase, but I guess we will encounter this problem for any block using __experimentalLabel, so I'd try to fix the root of the problem, not adding a workaround for the use cases we have identified.

Ok I've started looking at devising a general solution for this.

> For example, it seems it is also used in the image block: [link](https://github.com/WordPress/gutenberg/blob/322766e04ab3351df9cbf5ee22c220ff03a72d38/packages/block-library/src/image/index.js#L31). I haven't tested it, but maybe there is the same problem there.

Yes, the same problem exists with the image block, except that the image uses the `alt` attribute rather than a `content` attribute to create the label.

File: packages/block-editor/src/hooks/use-bindings-attributes.js [hunk 1]
```diff
@@ -296,9 +296,18 @@ export const withBlockBindingSupport = createHigherOrderComponent(
...
 		);
 	},
-	'withBlockBindingSupport'
+	'editWithBlockBindings'
 );
 
+function labelWithBlockBindings( settings ) {
+	return ( attributes, { context } ) => {
+		if ( attributes?.metadata?.bindings && context === 'accessibility' ) {
+			return 'Overriding label';
```
Comment (afercia) [inline packages/block-editor/src/hooks/use-bindings-attributes.js]: > For now, I just check whether the paragraph is bound and, if so, return the label `Paragraph with dynamic content`.

Is `Overriding label` a leftover added for testing? In any case, it should be translatable.

**Review by afercia — DISMISSED** [review]
Please see my last comment.

File: packages/block-editor/src/hooks/use-bindings-attributes.js (hunk 1 above)
Comment (artemiomorales) [reply packages/block-editor/src/hooks/use-bindings-attributes.js]: Thanks for the feedback! Rather than hardcoding this, I've now modified the PR so it reads the actual bindings values. I've updated the PR description accordingly.

**Comment by artemiomorales** [comment]
Update: I've revised the PR so that it retrieves the actual bindings value and uses it for the label. The PR description has been updated accordingly.

Note that pattern overrides required some additional handling. I explore a solution in https://github.com/WordPress/gutenberg/pull/65302, which is meant to accompany this PR or be merged shortly after.

**Comment by SantosGuillamot** [comment]
I didn't find time to properly review this yet (it's on my to-do list). But there are a couple of things I wanted to ask:
* Why do we need a different solution for pattern overrides? Is this going to be a problem for other registered sources? If that's the case, I wonder if we are solving the root of the problem.
* Did we explore the possibility of using the bindings value in `getBlockLabel` and `getBlockAccessibilityLabel`?

**Comment by artemiomorales** [comment]
@SantosGuillamot As far as I can tell, this is root of the problem: The block bindings hook in `use-bindings-attributes.js` overwrites the regular attributes with the bindings values in the `BlockEdit` component, and because the `BlockSelectionButton` exists at a different level of the hierarchy outside of the block edit context, the computed bindings information isn't easily available to it.

Here's where the attributes are overridden in the `BlockEdit` component:

https://github.com/WordPress/gutenberg/blob/b7c326c9f989452fd75ca5f6f55ad6fc031519fe/packages/block-editor/src/hooks/use-bindings-attributes.js#L304-L312

Worth noting is that `BlockEdit` is a context provider:

https://github.com/WordPress/gutenberg/blob/03d93b4624a42d795da71317ed731396e19acdd2/packages/block-editor/src/components/block-edit/index.js#L57-L99

Both `attributes` and `setAttributes` are passed to `BlockEdit` inside the BlockList component:

https://github.com/WordPress/gutenberg/blob/03d93b4624a42d795da71317ed731396e19acdd2/packages/block-editor/src/components/block-list/block.js#L130-L152

And the `BlockList` component is rendered inside of `BlockCanvas`:

https://github.com/WordPress/gutenberg/blob/de01da21b2b966ed58d11719e4fbc8b7bf114719/packages/block-editor/src/components/block-canvas/index.js#L26-L85

The `BlockSelectionButton` is part of `BlockTools`, which, as you can see in the snippet above, exists at a higher level than the `BlockList`. What we need to do is make the bound attributes available at that level.

To solve this, we could potentially refactor to wrap the `BlockTools` in another context provider, and use that to store the computed bindings values. Either that or refactor the `BlockEdit` component. That was my first thought, though it seems like it could get involved.

Another option, like I've done here, is to replicate the logic to compute the bindings values in the `BlockSelectionButton`. (This approach comes via a quick suggestion from @ellatrix — no need to join the conversation Ella, just pinging for awareness and in case you're interested in taking a look).

> Why do we need a different solution for pattern overrides? Is this going to be a problem for other registered sources? If that's the case, I wonder if we are solving the root of the problem.

It's not really a different solution — it's essentially replicating the logic that we normally use in  `use-bindings-attributes.js` to retrieve the overrides. There is a small detail with priority in how we're reading the custom name versus the block content for the label, but overall I just handled the patterns in a separate PR to make everything easier to parse.

As far as other registered sources goes, this issue with the label is currently specific to sources that override the `content` attribute in paragraphs and headings. As long as we pass the updated attributes computed via the normal bindings logic to `getBlockAccessibilityLabel`, then it should work as intended.

> Did we explore the possibility of using the bindings value in getBlockLabel and getBlockAccessibilityLabel?

`getBlockLabel` and `getBlockAccessibilityLabel` are utils that use the block attributes to create a label, so I don't think it's appropriate to extend them to read the bindings values. Instead, I think it makes sense to pass the right attributes to them so they work as intended.

So in summary, I see two solutions:
1. Somehow share the computed value of the bindings in `use-bindings-attributes.js` with `BlockTools` (which may require refactoring of where and how we define block context)
2. Just recompute the bindings values for the `BlockSelectionButton`

Perhaps there's an easy way to do the first approach, though I just wanted to get this working and keep iterating, and I didn't want to go down that potential rabbit hole unless we felt it was necessary. I could be off base here, though. Would be happy to hear any additional thoughts or insights 🙏

**Comment by SantosGuillamot** [comment]
I've been trying to test the original issue and consequently the potential fix, but it seems the navigation mode was removed in [this pull request](https://github.com/WordPress/gutenberg/pull/65204). More context can be found in [this other issue](https://github.com/WordPress/gutenberg/issues/65603), where a new mechanism is being discussed.

Having said that, I think this raises again the discussion about when to use the original block attributes and when to use the binding values. You can find more information in [this old pull request](https://github.com/WordPress/gutenberg/pull/61817), where that possibility was explored. My gut says we need to provide a way to let developers decide if they want to use the binding values or the original attributes, but the use cases weren't clear at that moment. This could be done via a new selector or even accepting a new parameter in `getBlockAttributes`. Anyway, that need to be explored and done carefully.

**Comment by artemiomorales** [comment]
> I've been trying to test the original issue and consequently the potential fix, but it seems the navigation mode was removed in https://github.com/WordPress/gutenberg/pull/65204. More context can be found in https://github.com/WordPress/gutenberg/issues/65603, where a new mechanism is being discussed.

Ok, on one hand, we could try to land this in core then, but any effort invested into a solution likely won't be viable for more than one release. I think we should see what the new mechanism ends up being before investing more time into this.

> My gut says we need to provide a way to let developers decide if they want to use the binding values or the original attributes, but the use cases weren't clear at that moment. This could be done via a new selector or even accepting a new parameter in getBlockAttributes. Anyway, that need to be explored and done carefully.

I think we could potentially offer the developers a way to customize what's used for the label, but what would the sensible default be? I think creating a label like, "{Name of attribute}: {Binding value}" could work well in the majority of cases. I can't say that for sure, though any approach we take would likely be better than the current one with post meta, where it just reads the original content.

**Comment by SantosGuillamot** [comment]
I may be wrong here but, if I remember this particular issue was only a problem for the Navigation mode, right? I think it was working fine in my tests in other modes.

When I said _"provide a way to let developers decide if they want to use the binding values or the original attributes"_ I wasn't referring to letting them decide the label discussed in this particular issue, I was referring to providing a method to let any developer, in any part of the code, decide if they want to use the original attributes or the values of the bindings.

**Comment by artemiomorales** [comment]
> I may be wrong here but, if I remember this particular issue was only a problem for the Navigation mode, right? I think it was working fine in my tests in other modes.

Right, since Navigation Move has been removed, then this PR is no longer relevant. Keyboard navigation no longer works well in the editor, but that's a bigger issue than just block bindings.

> When I said "provide a way to let developers decide if they want to use the binding values or the original attributes" I wasn't referring to letting them decide the label discussed in this particular issue, I was referring to providing a method to let any developer, in any part of the code, decide if they want to use the original attributes or the values of the bindings.

I think we're both in agreement here; I was just thinking of what a good default would be.

In any case, I'm going to close this. We can continue the conversation at a later date if it becomes relevant.
